File: flutter-src/test/schedule_test.dart
/// Dakar Bus - Groupe 11
/// Tests du socle horaires réels
/// Test A à G selon spécification

import 'package:test/test.dart';
import '../lib/models/data_status.dart';
import '../lib/models/departure.dart';
import '../lib/models/stop.dart';
import '../lib/repositories/unavailable_schedule_repository.dart';
import '../lib/services/schedule_service.dart' as service;

void main() {
  group('Groupe 11 - Socle horaires réels', () {
    test('Test A - Aucun repository réel : aucune source -> aucun départ -> unknown / indisponible', () async {
      final repo = UnavailableScheduleRepository();
      final departures = await repo.departuresForStop('TER_01_Dakar');
      expect(departures, isEmpty);

      final stop = Stop(id: 'TER_01_Dakar', name: 'Gare de Dakar', departures: departures);
      expect(stop.hasSourcedSchedule, isFalse);
      expect(stop.dataStatus, DataStatus.unknown);
      expect(stop.nextDepartureLabel(DateTime.utc(2026, 9, 22, 10, 0)), 'Horaire non disponible');
      expect(stop.remainingMinutes(DateTime.utc(2026, 9, 22, 10, 0)), isNull);
      expect(stop.departureAfter(DateTime.utc(2026, 9, 22, 10, 0)), isNull);
    });

    test('Test B - Départ scheduled réel injecté par le test', () {
      final fixedNow = DateTime.utc(2026, 9, 22, 13, 0);
      final departureTime = DateTime.utc(2026, 9, 22, 14, 20);

      final dep = Departure(
        stopId: 'TER_01_Dakar',
        lineId: 'TER',
        departureTime: departureTime,
        status: DataStatus.scheduled,
        direction: 'Diamniadio',
        sourceId: 'test-gtfs-2026',
      );

      expect(dep.status, DataStatus.scheduled);

      final stop = Stop(id: 'TER_01_Dakar', name: 'Gare de Dakar', departures: [dep]);
      expect(stop.hasSourcedSchedule, isTrue);

      final next = stop.departureAfter(fixedNow);
      expect(next, isNotNull);
      expect(next!.departureTime, departureTime);
      expect(next.status, DataStatus.scheduled);
      expect(stop.nextDepartureLabel(fixedNow), isNot('Horaire non disponible'));
      expect(stop.remainingMinutes(fixedNow), 80);
    });

    test('Test C - Départ live réel injecté', () {
      final fixedNow = DateTime.utc(2026, 9, 22, 10, 0);
      final departureTime = DateTime.utc(2026, 9, 22, 10, 5);

      final dep = Departure(
        stopId: 'BRT_01_Petersen',
        lineId: 'BRT 01',
        departureTime: departureTime,
        status: DataStatus.live,
        direction: 'Guédiawaye',
        sourceId: 'test-realtime-cetud',
      );

      final stop = Stop(id: 'BRT_01_Petersen', name: 'Petersen', departures: [dep]);
      final next = stop.departureAfter(fixedNow);
      expect(next, isNotNull);
      expect(next!.status, DataStatus.live);
      expect(stop.remainingMinutes(fixedNow), 5);
      expect(stop.nextDepartureLabel(fixedNow), isNot('Horaire non disponible'));
    });

    test('Test D - Départ passé ne doit pas être retourné', () {
      final now = DateTime.utc(2026, 9, 22, 15, 0);
      final past = DateTime.utc(2026, 9, 22, 14, 0);

      final dep = Departure(
        stopId: 'TER_01_Dakar',
        lineId: 'TER',
        departureTime: past,
        status: DataStatus.scheduled,
        sourceId: 'test',
      );

      final stop = Stop(id: 'TER_01_Dakar', name: 'Gare de Dakar', departures: [dep]);
      expect(stop.departureAfter(now), isNull);
      expect(stop.remainingMinutes(now), isNull);
      expect(stop.nextDepartureLabel(now), 'Horaire non disponible');
    });

    test('Test E - Plusieurs départs 13:00, 13:30, 14:00', () {
      final d1 = Departure(stopId: 'S1', lineId: 'TER', departureTime: DateTime.utc(2026, 9, 22, 13, 0), status: DataStatus.scheduled, sourceId: 'test');
      final d2 = Departure(stopId: 'S1', lineId: 'TER', departureTime: DateTime.utc(2026, 9, 22, 13, 30), status: DataStatus.scheduled, sourceId: 'test');
      final d3 = Departure(stopId: 'S1', lineId: 'TER', departureTime: DateTime.utc(2026, 9, 22, 14, 0), status: DataStatus.scheduled, sourceId: 'test');

      final stop = Stop(id: 'S1', name: 'Test', departures: [d1, d2, d3]);

      expect(stop.departureAfter(DateTime.utc(2026, 9, 22, 12, 50))!.departureTime, DateTime.utc(2026, 9, 22, 13, 0));
      expect(stop.departureAfter(DateTime.utc(2026, 9, 22, 13, 15))!.departureTime, DateTime.utc(2026, 9, 22, 13, 30));
      expect(stop.departureAfter(DateTime.utc(2026, 9, 22, 13, 45))!.departureTime, DateTime.utc(2026, 9, 22, 14, 0));
      expect(stop.departureAfter(DateTime.utc(2026, 9, 22, 14, 10)), isNull);
    });

    test('Test F - Aucun départ : null et Horaire non disponible', () {
      final stop = Stop(id: 'EMPTY', name: 'Vide', departures: []);
      final now = DateTime.utc(2026, 9, 22, 10, 0);
      expect(stop.departureAfter(now), isNull);
      expect(stop.remainingMinutes(now), isNull);
      expect(stop.nextDepartureLabel(now), 'Horaire non disponible');
      expect(stop.hasSourcedSchedule, isFalse);
      expect(stop.dataStatus, DataStatus.unknown);
    });

    test('Test G - Pas de génération automatique', () async {
      final repo = UnavailableScheduleRepository();
      for (final id in ['TER_01_Dakar', 'BRT_01_Petersen', 'UNKNOWN_STOP']) {
        final deps = await repo.departuresForStop(id);
        expect(deps, isEmpty);
        final stop = Stop(id: id, name: id, departures: deps);
        expect(stop.hasSourcedSchedule, isFalse);
        expect(stop.departures, isEmpty);
      }
      final emptyStop = Stop(id: 'TEST', name: 'Test');
      expect(emptyStop.departures, isEmpty);
      expect(emptyStop.generateSchedule(), isEmpty);
    });

    test('Test services avec temps contrôlable', () {
      final now = DateTime.utc(2026, 9, 22, 8, 0);
      final deps = [
        Departure(stopId: 'S1', lineId: 'L1', departureTime: DateTime.utc(2026, 9, 22, 9, 0), status: DataStatus.scheduled, sourceId: 'test'),
        Departure(stopId: 'S1', lineId: 'L1', departureTime: DateTime.utc(2026, 9, 22, 10, 0), status: DataStatus.live, sourceId: 'test'),
      ];

      final next = service.departureAfter(deps, now);
      expect(next!.departureTime, DateTime.utc(2026, 9, 22, 9, 0));
      expect(service.remainingMinutes(deps, now), 60);
      expect(service.nextDepartureLabel(deps, now), isNot('Horaire non disponible'));
      expect(service.getDataStatus(deps, now), DataStatus.scheduled);

      final afterFirst = DateTime.utc(2026, 9, 22, 9, 30);
      final next2 = service.departureAfter(deps, afterFirst);
      expect(next2!.status, DataStatus.live);
      expect(service.getDataStatus(deps, afterFirst), DataStatus.live);
    });
  });
}

File: tests/schedule.test.js
'use strict';
const { test } = require('node:test');
const assert = require('node:assert/strict');

// Import ES modules via dynamic import workaround for CommonJS test runner
// We'll use require for built files via creating CJS versions or using import()

// Since our implementation is ESM, we need to create a CJS compatible test harness
// We'll directly test logic by importing via import() in async tests

test('Test A — Aucun repository réel : aucune source → aucun départ → unknown / indisponible', async () => {
  const { UnavailableScheduleRepository } = await import('../js/repositories/UnavailableScheduleRepository.js');
  const { Stop } = await import('../js/models/Stop.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const repo = new UnavailableScheduleRepository();
  const departures = await repo.departuresForStop('TER_01_Dakar');
  assert.deepEqual(departures, [], 'Unavailable repository must return empty list');

  const stop = new Stop({ id: 'TER_01_Dakar', name: 'Gare de Dakar', departures });
  assert.equal(stop.hasSourcedSchedule, false);
  assert.equal(stop.dataStatus, DataStatus.unknown);
  assert.equal(stop.nextDepartureLabel(new Date('2026-09-22T10:00:00Z')), 'Horaire non disponible');
  assert.equal(stop.remainingMinutes(new Date('2026-09-22T10:00:00Z')), null);
  assert.equal(stop.departureAfter(new Date('2026-09-22T10:00:00Z')), null);
});

test('Test B — Départ scheduled réel injecté par le test', async () => {
  const { Departure } = await import('../js/models/Departure.js');
  const { Stop } = await import('../js/models/Stop.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const fixedNow = new Date('2026-09-22T13:00:00Z');
  const departureTime = new Date('2026-09-22T14:20:00Z');

  const dep = new Departure({
    stopId: 'TER_01_Dakar',
    lineId: 'TER',
    departureTime,
    status: DataStatus.scheduled,
    direction: 'Diamniadio',
    sourceId: 'test-gtfs-2026'
  });

  assert.equal(dep.status, DataStatus.scheduled);
  assert.equal(dep.stopId, 'TER_01_Dakar');

  const stop = new Stop({ id: 'TER_01_Dakar', name: 'Gare de Dakar', departures: [dep] });
  assert.equal(stop.hasSourcedSchedule, true);

  const next = stop.departureAfter(fixedNow);
  assert.ok(next, 'should have next departure');
  assert.equal(next.departureTime.getTime(), departureTime.getTime());
  assert.equal(next.status, DataStatus.scheduled);

  // Label should be real time, not "Horaire non disponible"
  const label = stop.nextDepartureLabel(fixedNow);
  // Should contain "14 h 20" in local time, but our Date is UTC, getHours uses local TZ.
  // Use UTC-based check: we format with getHours, so we check that label is not unavailable and contains "h"
  assert.notEqual(label, 'Horaire non disponible');
  assert.ok(label.includes('h'), `label should contain h, got ${label}`);

  // remainingMinutes should be 80
  const remaining = stop.remainingMinutes(fixedNow);
  assert.equal(remaining, 80);
});

test('Test C — Départ live réel injecté par le test', async () => {
  const { Departure } = await import('../js/models/Departure.js');
  const { Stop } = await import('../js/models/Stop.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const fixedNow = new Date('2026-09-22T10:00:00Z');
  const departureTime = new Date('2026-09-22T10:05:00Z');

  const dep = new Departure({
    stopId: 'BRT_01_Petersen',
    lineId: 'BRT 01',
    departureTime,
    status: DataStatus.live,
    direction: 'Guédiawaye',
    sourceId: 'test-realtime-cetud'
  });

  const stop = new Stop({ id: 'BRT_01_Petersen', name: 'Petersen', departures: [dep] });

  const next = stop.departureAfter(fixedNow);
  assert.ok(next);
  assert.equal(next.status, DataStatus.live);
  assert.equal(next.departureTime.getTime(), departureTime.getTime());

  assert.equal(stop.remainingMinutes(fixedNow), 5);
  assert.notEqual(stop.nextDepartureLabel(fixedNow), 'Horaire non disponible');
});

test('Test D — Départ passé ne doit pas être retourné comme prochain départ', async () => {
  const { Departure } = await import('../js/models/Departure.js');
  const { Stop } = await import('../js/models/Stop.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const now = new Date('2026-09-22T15:00:00Z');
  const past = new Date('2026-09-22T14:00:00Z');

  const dep = new Departure({
    stopId: 'TER_01_Dakar',
    lineId: 'TER',
    departureTime: past,
    status: DataStatus.scheduled,
    sourceId: 'test'
  });

  const stop = new Stop({ id: 'TER_01_Dakar', name: 'Gare de Dakar', departures: [dep] });

  assert.equal(stop.departureAfter(now), null);
  assert.equal(stop.remainingMinutes(now), null);
  assert.equal(stop.nextDepartureLabel(now), 'Horaire non disponible');
});

test('Test E — Plusieurs départs : 13:00, 13:30, 14:00', async () => {
  const { Departure } = await import('../js/models/Departure.js');
  const { Stop } = await import('../js/models/Stop.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const d1 = new Departure({ stopId: 'S1', lineId: 'TER', departureTime: new Date('2026-09-22T13:00:00Z'), status: DataStatus.scheduled, sourceId: 'test' });
  const d2 = new Departure({ stopId: 'S1', lineId: 'TER', departureTime: new Date('2026-09-22T13:30:00Z'), status: DataStatus.scheduled, sourceId: 'test' });
  const d3 = new Departure({ stopId: 'S1', lineId: 'TER', departureTime: new Date('2026-09-22T14:00:00Z'), status: DataStatus.scheduled, sourceId: 'test' });

  const stop = new Stop({ id: 'S1', name: 'Test', departures: [d1, d2, d3] });

  // Avant 13:00 -> 13:00
  let now = new Date('2026-09-22T12:50:00Z');
  let next = stop.departureAfter(now);
  assert.equal(next.departureTime.toISOString(), '2026-09-22T13:00:00.000Z');

  // Entre 13:00 et 13:30 -> 13:30
  now = new Date('2026-09-22T13:15:00Z');
  next = stop.departureAfter(now);
  assert.equal(next.departureTime.toISOString(), '2026-09-22T13:30:00.000Z');

  // Entre 13:30 et 14:00 -> 14:00
  now = new Date('2026-09-22T13:45:00Z');
  next = stop.departureAfter(now);
  assert.equal(next.departureTime.toISOString(), '2026-09-22T14:00:00.000Z');

  // Après 14:00 -> null
  now = new Date('2026-09-22T14:10:00Z');
  next = stop.departureAfter(now);
  assert.equal(next, null);
});

test('Test F — Aucun départ : null et Horaire non disponible', async () => {
  const { Stop } = await import('../js/models/Stop.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const stop = new Stop({ id: 'EMPTY', name: 'Vide', departures: [] });
  const now = new Date('2026-09-22T10:00:00Z');

  assert.equal(stop.departureAfter(now), null);
  assert.equal(stop.remainingMinutes(now), null);
  assert.equal(stop.nextDepartureLabel(now), 'Horaire non disponible');
  assert.equal(stop.hasSourcedSchedule, false);
  assert.equal(stop.dataStatus, DataStatus.unknown);
});

test('Test G — Pas de génération automatique : prouver qu\'aucun horaire n\'est créé sans source', async () => {
  const { UnavailableScheduleRepository } = await import('../js/repositories/UnavailableScheduleRepository.js');
  const { Stop } = await import('../js/models/Stop.js');

  const repo = new UnavailableScheduleRepository();

  // Simuler plusieurs appels
  const ids = ['TER_01_Dakar', 'BRT_01_Petersen', 'UNKNOWN_STOP'];
  for (const id of ids) {
    const deps = await repo.departuresForStop(id);
    assert.deepEqual(deps, [], `Repository should return empty for ${id}, not generate`);
    assert.equal(deps.length, 0);

    const stop = new Stop({ id, name: id, departures: deps });
    assert.equal(stop.hasSourcedSchedule, false);
    assert.equal(stop.departures.length, 0);
  }

  // Vérifier que Stop sans departures ne génère rien en interne
  const emptyStop = new Stop({ id: 'TEST', name: 'Test' });
  assert.equal(emptyStop.departures.length, 0);
  assert.equal(emptyStop._generateSchedule().length, 0, '_generateSchedule should return empty (deprecated)');
});

test('Test non-régression — DataStatus et DataTrust distincts', async () => {
  const { DataStatus } = await import('../js/models/DataStatus.js');
  const { DataTrust } = await import('../js/models/DataTrust.js');

  // DataStatus
  assert.equal(DataStatus.scheduled, 'scheduled');
  assert.equal(DataStatus.live, 'live');
  assert.equal(DataStatus.unknown, 'unknown');

  // DataTrust différent conceptuellement, même si string 'unknown' peut coïncider
  assert.equal(DataTrust.verified, 'verified');
  assert.notEqual(DataStatus.scheduled, DataTrust.verified);
  // Les deux enums sont des objets différents, pas fusionnés
  assert.notDeepEqual(DataStatus, DataTrust);
  // DataTrust décrit confiance, DataStatus disponibilité temporelle
  assert.ok(DataStatus.scheduled);
  assert.ok(DataTrust.verified);
  // Vérifie que les clés sont distinctes
  assert.ok(Object.keys(DataStatus).includes('scheduled'));
  assert.ok(Object.keys(DataTrust).includes('verified'));
});

test('Test Departure immutability et validation', async () => {
  const { Departure } = await import('../js/models/Departure.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const dep = new Departure({
    stopId: 'S1',
    lineId: 'L1',
    departureTime: new Date('2026-09-22T12:00:00Z'),
    status: DataStatus.scheduled,
  });

  // Immutable
  assert.ok(Object.isFrozen(dep));

  // Invalid should throw
  assert.throws(() => {
    new Departure({ stopId: '', lineId: 'L1', departureTime: new Date(), status: DataStatus.scheduled });
  });

  assert.throws(() => {
    new Departure({ stopId: 'S1', lineId: 'L1', departureTime: new Date('invalid'), status: DataStatus.scheduled });
  });

  assert.throws(() => {
    new Departure({ stopId: 'S1', lineId: 'L1', departureTime: new Date(), status: 'fake' });
  });
});

test('Test service functions avec temps contrôlable', async () => {
  const { departureAfter, remainingMinutes, nextDepartureLabel, getDataStatus } = await import('../js/services/scheduleService.js');
  const { Departure } = await import('../js/models/Departure.js');
  const { DataStatus } = await import('../js/models/DataStatus.js');

  const now = new Date('2026-09-22T08:00:00Z');
  const deps = [
    new Departure({ stopId: 'S1', lineId: 'L1', departureTime: new Date('2026-09-22T09:00:00Z'), status: DataStatus.scheduled, sourceId: 'test' }),
    new Departure({ stopId: 'S1', lineId: 'L1', departureTime: new Date('2026-09-22T10:00:00Z'), status: DataStatus.live, sourceId: 'test' }),
  ];

  const next = departureAfter(deps, now);
  assert.equal(next.departureTime.toISOString(), '2026-09-22T09:00:00.000Z');
  assert.equal(remainingMinutes(deps, now), 60);
  assert.notEqual(nextDepartureLabel(deps, now), 'Horaire non disponible');
  assert.equal(getDataStatus(deps, now), DataStatus.scheduled);

  const afterFirst = new Date('2026-09-22T09:30:00Z');
  const next2 = departureAfter(deps, afterFirst);
  assert.equal(next2.status, DataStatus.live);
  assert.equal(getDataStatus(deps, afterFirst), DataStatus.live);
});
